features/address.js

```diff
@@ -0,0 +1,252 @@
+  }
+
+  let addressPayload, addressAttributes
+  if (derivationScheme === 1 && derivationPath.length > 0) {
+    addressPayload = encryptDerivationPath(derivationPath, hdPassphrase)
+    addressAttributes = new Map([[1, cbor.encode(addressPayload)]])
+  } else {
+    addressPayload = Buffer.from([])
+    addressAttributes = new Map()
+  }
+
+  if (protocolMagic !== MAINNET_PROTOCOL_MAGIC) {
+    addressAttributes.set(2, cbor.encode(protocolMagic))
+  }
+
+  const getAddressRootHash = (input) => blake2b(sha3_256(cbor.encode(input)), 28)
+
+  const addressRoot = getAddressRootHash([
+    0,
+    [0, xpub],
+    addressPayload.length > 0 ? new Map([[1, cbor.encode(addressPayload)]]) : new Map(),
+  ])
+  const addressType = 0 // Public key address
+  const addressData = [addressRoot, addressAttributes, addressType]
+  const addressDataEncoded = cbor.encode(addressData)
+
+  return cbor.encode([new cbor.Tagged(24, addressDataEncoded), crc32(addressDataEncoded)])
+}
+
+function getAddressHeader(addressType, networkId) {
+  return Buffer.from([(addressType << 4) | networkId])
+}
+
+function getPubKeyBlake2b224Hash(pubKey) {
+  validateBuffer(pubKey, PUB_KEY_LEN)
+
+  return blake2b(pubKey, KEY_HASH_LEN)
+}
+
+function packBaseAddress(spendingKeyHash, stakingKeyHash, networkId) {
+  validateBuffer(spendingKeyHash, KEY_HASH_LEN)
+  validateBuffer(stakingKeyHash, KEY_HASH_LEN)
+  validateNetworkId(networkId)
+
+  return Buffer.concat([
+    getAddressHeader(AddressTypes.BASE, networkId),
+    spendingKeyHash,
+    stakingKeyHash,
+  ])
+}
+
+function packPointerAddress(pubKeyHash, pointer, networkId) {
+  validateBuffer(pubKeyHash, KEY_HASH_LEN)
+  validatePointer(pointer)
+  validateNetworkId(networkId)
+
+  const {blockIndex, txIndex, certificateIndex} = pointer
+
+  return Buffer.concat([
+    getAddressHeader(AddressTypes.POINTER, networkId),
+    pubKeyHash,
+    Buffer.concat([
+      variableLengthEncode(blockIndex),
+      variableLengthEncode(txIndex),
+      variableLengthEncode(certificateIndex)
+    ])
+  ])
+}
+
+function packEnterpriseAddress(spendingKeyHash, networkId) {
+  validateBuffer(spendingKeyHash, KEY_HASH_LEN)
+  validateNetworkId(networkId)
+
+  return Buffer.concat([
+    getAddressHeader(AddressTypes.ENTERPRISE, networkId),
+    spendingKeyHash
+  ])
+}
+
+function packRewardAddress(stakingKeyHash, networkId) {
+  validateBuffer(stakingKeyHash, KEY_HASH_LEN)
+  validateNetworkId(networkId)
+
+  return Buffer.concat([
+    getAddressHeader(AddressTypes.REWARD, networkId),
+    stakingKeyHash
+  ])
+}
+
+function unpackBootstrapAddress(address, hdPassphrase) {
+  // we decode the address from the base58 string
+  // and then we strip the 24 CBOR data tags (the "[0].value" part)
+  const addressAsBuffer = cbor.decode(base58.decode(address))[0].value
+  const addressData = cbor.decode(addressAsBuffer)
+  const addressAttributes = addressData[1]
+  const addressPayload = cbor.decode(addressAttributes.get(1))
```

Comment:
@refi93  There seems to be a bug. First of all `addressAttributes.get(1)` throws an error, since `addressAttributes` isn't Map (at least not in case its empty)
If anything it should be something like this 
`const addressPayload = addressAttributes[1] ? cbor.decode(addressAttributes.get[1]) : something`. I am really not sure what to set as addressPayload in case its not in addressAtrributes.

Reply:
good catch @PeterBenc - this function would work only for V1 (Daedalus addresses) which have non-empty attributes. Moreover, cbor lib seems to be decoding empty map to empty object which does not have the "get" attribute.

I gave it a second thought and the `unpackBootstrapAddress` call is imho misleading the way it currently works as its output is not the inverse of `packBootstrapAddress` which takes more parameters. So I decided to break the function down into `getBootstrapAddressAttributes`, `getBootstrapAddressDerivationPath` and `getBootstrapAddressProtocolMagic` and I refoctored the api to expect the addresses as buffers to stay consistent with similar calls for Shelley addresses. And I added the addressToBuffer utility function which translates string representation of addresses to buffers.